src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,218 @@
+    {
+        protected override IRemoteAutoInsertService CreateService(in ServiceArgs args)
+            => new RemoteAutoInsertService(in args);
+    }
+
+    private readonly IAutoInsertService _autoInsertService
+        = args.ExportProvider.GetExportedValue<IAutoInsertService>();
+    private readonly IDocumentMappingService _documentMappingService
+        = args.ExportProvider.GetExportedValue<IDocumentMappingService>();
+    private readonly IFilePathService _filePathService =
+        args.ExportProvider.GetExportedValue<IFilePathService>();
+    private readonly IRazorFormattingService _razorFormattingService =
+        args.ExportProvider.GetExportedValue<IRazorFormattingService>();
+    private readonly ILogger _logger =
+        args.ExportProvider.GetExportedValue<ILoggerFactory>()
+        .GetOrCreateLogger(nameof(RemoteAutoInsertService));
+
+    public ValueTask<Response> TryResolveInsertionAsync(
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        bool formatOnType,
+        bool indentWithTabs,
+        int indentSize,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => TryResolveInsertionAsync(
+                context,
+                linePosition,
+                character,
+                autoCloseTags,
+                formatOnType,
+                indentWithTabs,
+                indentSize,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> TryResolveInsertionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        bool formatOnType,
+        bool indentWithTabs,
+        int indentSize,
+        CancellationToken cancellationToken)
+    {
+        var sourceText = await remoteDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var index))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var codeDocument = await remoteDocumentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        // Always try our own service first, regardless of language
+        // E.g. if ">" is typed for html tag, it's actually our auto-insert provider
+        // that adds closing tag instead of HTML even though we are in HTML
+        var insertTextEdit = _autoInsertService.TryResolveInsertion(
+            codeDocument,
+            VsLspExtensions.ToPosition(linePosition),
+            character,
+            autoCloseTags);
+
+        if (insertTextEdit is { } edit)
+        {
+            return Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(edit));
+        }
```

Comment:
After transforming `TryResultInsertion` to a `bool` and `out` parameter, you can rewrite this like so:

```suggestion
        if (_autoInsertService.TryResolveInsertion(codeDocument, VsLspExtensions.ToPosition(linePosition), character, autoCloseTags, out var insertTextEdit))
        {
            return Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(edit));
        }
```

Reply:
Without converting to bool, I could've written this as 

if(_autoInsertService.TryResolveInsertion(
            codeDocument,
            VsLspExtensions.ToPosition(linePosition),
            character,
            autoCloseTags) is {  } edit)
{
            return Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(edit));
        }
}

right? The mutli-statement form seemed more readable to me.

But I did change to bool per PR request from both of you, and will change this as well of course.

Reply:
Sorry, still not at all convinced that using bool and out is any more succinct :grinning: , but happy to go with the majority of the team of course.  :blush:

Reply:
> The mutli-statement form seemed more readable to me.

We will just have to agree to disagree. I definitely do not prefer it when a single line contains a single variable name that is unlikely to be important in isolation.